core/src/view/view.cpp

```diff
@@ -128,23 +152,24 @@ glm::dmat2 View::getBoundsRect() const {
 
 }
 
-float View::toWorldDistance(float _screenDistance) const {
-    float metersPerTile = 2 * MapProjection::HALF_CIRCUMFERENCE * pow(2, -m_zoom);
-    return _screenDistance * metersPerTile / (m_pixelScale * m_pixelsPerTile);
-}
-
-void View::toWorldDisplacement(float& _screenX, float& _screenY) const {
+void View::screenToGroundPlane(float& _screenX, float& _screenY) const {
+    
+    // Cast a ray and find its intersection with the z = 0 plane,
+    // following the technique described here: http://antongerdelan.net/opengl/raycasting.html
     
-    float metersPerPixel = toWorldDistance(1);
+    glm::vec4 ray_clip = { 2.f * _screenX / m_vpWidth - 1.f, 1.f - 2.f * _screenY / m_vpHeight, -1.f, 1.f }; // Ray from camera in clip space
+    glm::vec3 ray_world = glm::vec3(m_invViewProj * ray_clip); // Ray from camera in world space
     
-    // Rotate screen displacement into world space
-    float cos_r = cos(m_roll);
-    float sin_r = sin(m_roll);
-    float x = (_screenX * cos_r + _screenY * sin_r) * metersPerPixel;
-    float y = (_screenX * -sin_r + _screenY * cos_r) * metersPerPixel;
+    float t; // Distance along ray to ground plane
+    if (ray_world.z != 0.f) {
+        t = -m_pos.z / ray_world.z;
+    } else {
+        t = 0;
+    }
     
-    _screenX = x;
-    _screenY = y;
+    ray_world *= t;
```

Comment:
I know this will never happen considering we are restricting how much we can shove!!... but should we put in a check for t being positive? Negative t values mean ray will never intersect the ground plane (which is when world is upside down :D)